lib/susedistribution.pm

```diff
@@ -173,13 +173,18 @@ sub init_desktop_runner {
         } else {
             type_string $program;
         }
+        # Do multiple attempts only on plasma
+        last unless (check_var('DESKTOP', 'kde'));
         # Make sure we have plasma suggestions as it may take time, especially on boot or under load. Otherwise, try again
-        last unless (check_var('DESKTOP', 'kde') && !check_screen('desktop-runner-plasma-suggestions'));
-        if ($retries > 1) {
+        if ($retries == 1) {
+            assert_screen('desktop-runner-plasma-suggestions', $timeout);
+        } elsif (!check_screen('desktop-runner-plasma-suggestions', $timeout)) {
```

Comment:
I consider this approach racy as well. Can you evaluate if you can use a multi-tag `assert_screen` with `match_has_tag` instead of `check_screen` with non-zero timeout to prevent introducing any timing dependant behaviour, to save test execution time as well as state more explicitly from the testers point of view what are the expected alternatives. For example:

```perl
assert_screen([qw(yast2_console-finished yast2_missing_package)]);
if (match_has_tag('yast2_missing_package')) {
    send_key 'alt-o';  # confirm package installation
    assert_screen 'yast2_console-finished';
}
```

I agree that the previous approach by @ggardet might also not be the perfect solution in the end. Please see https://progress.opensuse.org/issues/35589 with my – rather lengthy – investigation. And any try to fix this proberly demands good statistics for a proof :) If you like you can actually use ressources on openqa.opensuse.org yourself for testing, @ggardet does the same with good success already. http://open.qa/docs/#_triggering_tests_based_on_an_any_remote_git_refspec_or_open_github_pull_request describes how that can be achieved. Feel free to ping "okurz" on irc://chat.freenode.net/opensuse-factory for help and getting API access to the instance.

Reply:
> I consider this approach racy as well. Can you evaluate if you can use a multi-tag `assert_screen` with `match_has_tag` instead of `check_screen` with non-zero timeout to prevent introducing any timing dependant behaviour, to save test execution time as well as state more explicitly from the testers point of view what are the expected alternatives.

Can you please elaborate which events would lead to a race here? Obviously, krunner needs **some** time to show the results, so checking immediately after the last character typed has a very low success rate.
The check here is only for the krunner suggestions, so there is exactly **one tag** involved.

Reply:
@okurz ^

Reply:
ok, sorry. in this case I don't mean "racy" but wasting time in case of no-match. In this specific case I actually think `check_screen` with a non-zero timeout would work ok but unfortunately I had been repeatedly fighting against many `check_screen` calls which are wasting time a lot and nobody realizes. I would appreciate if you can try to fix the same problem with the multi-tag `assert_screen` I suggested above. If you want to keep the `check_screen` call still I will accept that and merge the PR though :)

Reply:
The "nobody realizes" can not happen here, as the final repetition uses an "assert_screen", see https://openqa.opensuse.org/tests/962535#step/dolphin/46 (that one was still missing a matching needle).
Anything but check_screen would likely waste more time:

1. with "0" timeout:
- the command is typed, krunner starts processing
- screen is grabbed, check_screen fails
- wait some time?
- check again?

2. with short timeout
- the command is typed, krunner starts processing
- screen is grabbed, check_screen does not match, but no timeout
- the suggestions appear
- screen is grabbed, check_screen succeeds immediately

An `assert_screen [ 'desktop-runner-plasma-suggestions', 'desktop-runner' ]` would immeditaly succeed even without the suggestions, so we would have to repeat it - but immediately, or after waiting, and how often? Imagine the suggestions appear only after 5 seconds (which does not seem to be the case here, but after 1 frame at most) ...

Reply:
Well `assert_screen ['desktop-runner-plasma-suggestions', 'desktop-runner' ]` is obviously not the right call on plasma but we can agree that it simply has to show the suggestions before we should try to press return. IMHO 9e0534af4 was already wrong because it introduced a check for the suggestions when the task of "init_desktop_runner" should just be to type, nothing more. We are looking *again* for the suggestions window in `x11_start_program` where we also try to handle retries. Your PR here seems to make things better but I actually tend towards reverting 9e0534af4 as well. WDYT? @ggardet as well?

Reply:
@okurz The thing is https://github.com/os-autoinst/os-autoinst-distri-opensuse/commit/9e0534af4b6478138d3c0a79215ccc2468ce1ad9 make things to work. Without https://github.com/os-autoinst/os-autoinst-distri-opensuse/commit/9e0534af4b6478138d3c0a79215ccc2468ce1ad9 there are a lots of failure on aarch64 (also seen on x86_64, with less occurrences). So, I would prefer to merge this current PR which improve things a lot. We still can find another working solution afterwards. But please do not revert PR which make things to work without another PR which fixes previously fixed/workarounded problems.

Reply:
convinced :)